In [1]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [2]:
df=pd.read_csv('C:/Users/Sohail/Desktop/fakenewscheck.csv')

In [3]:
df.head(10)

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   
5        6903                                        Tehran, USA   
6        7341  Girl Horrified At What She Watches Boyfriend D...   
7          95                  ‘Britain’s Schindler’ Dies at 106   
8        4869  Fact check: Trump and Clinton at the 'commande...   
9        2909  Iran reportedly makes new push for uranium con...   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL  
5    \nI’m not an immigrant, but my grandparents ...  FAKE  
6  Share This Baylee Luciani (left), Screenshot o...  FAKE  
7  A Czech stockbroker who saved more than 650 Je...  REAL  
8  Hillary Clinton and Donald Trump made some ina...  REAL  
9  Iranian negotiators reportedly have made a las...  REAL

In [4]:
labelling=df.label
labelling.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

### 1.  Using TfidVectorizer and Passive Aggressive Classifier

In [5]:
x_train,x_test,y_train,y_test=train_test_split(df['text'], labelling, test_size=0.3, random_state=10)

In [6]:
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.6)
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

In [8]:
packing=PassiveAggressiveClassifier(max_iter=45)
packing.fit(tfidf_train,y_train)
y_pred=packing.predict(tfidf_test)
auc_score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(auc_score*100,2)}%')

Accuracy: 93.9%


### Here the accuracy score of the Passive Aggressive Classifier is 93.9% ~ 94% which is considered to be highly accurate so we can use this model on our dataset

In [10]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[878,  42],
       [ 74, 907]], dtype=int64)

### The Number of true positive for our model is 878, number of true negative is 907,number of false positive is 74 and false negative is 42

### 2.  Using Count Vectorizer and Multinomial NB

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
test = MultinomialNB()

In [16]:
from sklearn.feature_extraction.text import CountVectorizer
count_vectorizer = CountVectorizer(stop_words='english')

In [18]:
count_train = count_vectorizer.fit_transform(x_train) 

In [19]:
count_test = count_vectorizer.transform(x_test)

In [20]:
count_df = pd.DataFrame(count_train.A, columns=count_vectorizer.get_feature_names())

In [21]:
test.fit(count_train, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [22]:
prediction = test.predict(count_test)

In [28]:
import sklearn.metrics as metrics
auc_score_count = metrics.accuracy_score(y_test, prediction)
print(f'Accuracy: {round(auc_score_count*100,2)}%')

Accuracy: 88.85%


### Here the accuracy score of the Multinomial Naive Bayes Classifer is 88.85% which is considered less than than of Passive Aggressive Classifer.So, Passive Aggressive classifer proves to be more suitable for this dataset than the significant other